components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/dao/SAMLSSOServiceProviderDAOTest.java

```diff
@@ -0,0 +1,470 @@
+        assertEquals(captor.getValue(), expectedPath, "Resource is not added at correct path");
+    }
+
+    @Test
+    public void testAddExistingServiceProvider() throws Exception {
+        SAMLSSOServiceProviderDO serviceProviderDO = new SAMLSSOServiceProviderDO();
+        String existingPath = getPath("existingIssuer");
+        serviceProviderDO.setIssuer("existingIssuer");
+        when(mockRegistry.resourceExists(existingPath)).thenReturn(true);
+        assertFalse(objUnderTest.addServiceProvider(serviceProviderDO), "Resource should not have added.");
+    }
+
+    @Test
+    public void testAddServiceProviderRegistryError() throws Exception {
+        SAMLSSOServiceProviderDO serviceProviderDO = new SAMLSSOServiceProviderDO();
+        String existingPath = getPath("erringIssuer");
+        serviceProviderDO.setIssuer("erringIssuer");
+        doThrow(RegistryException.class).when(mockRegistry).put(eq(existingPath), any(Resource.class));
+        try {
+            objUnderTest.addServiceProvider(serviceProviderDO);
+        } catch (IdentityException e) {
+            assertEquals(e.getMessage(), "Error while adding Service Provider", "Invalid error thrown for SP registry" +
+                    " add operation");
+        }
+    }
+
+    @Test
+    public void testGetServiceProviders() throws Exception {
+        when(mockRegistry.resourceExists(IdentityRegistryResources.SAML_SSO_SERVICE_PROVIDERS)).thenReturn(true);
+        Resource collection = new CollectionImpl();
+        String[] paths = new String[]{
+                getPath("DummyIssuer"), getPath("DummyAdvIssuer")
+        };
+        Properties dummyResourceProperties = new Properties();
+        dummyResourceProperties.putAll(dummyBasicProperties);
+        Resource dummyResource = new ResourceImpl();
+        dummyResource.setProperties(dummyResourceProperties);
+
+        Properties dummyAdvProperties = new Properties();
+        dummyAdvProperties.putAll((Map<?, ?>) dummyAdvProperties);
+        Resource dummyAdvResource = new ResourceImpl();
+        dummyAdvResource.setProperties(dummyAdvProperties);
+        Resource[] spResources = new Resource[]{dummyResource, dummyAdvResource};
+        collection.setContent(paths);
+        when(mockRegistry.get(IdentityRegistryResources.SAML_SSO_SERVICE_PROVIDERS)).thenReturn(collection);
+        when(mockRegistry.get(paths[0])).thenReturn(spResources[0]);
+        when(mockRegistry.get(paths[1])).thenReturn(spResources[1]);
+        SAMLSSOServiceProviderDO[] serviceProviders = objUnderTest.getServiceProviders();
+        assertEquals(serviceProviders.length, 2, "Should have returned 2 service providers.");
+    }
+
+    @Test
+    public void testRemoveServiceProvider() throws Exception {
+        String existingIssuer = "ExistingIssuer";
+        String path = getPath(existingIssuer);
+        when(mockRegistry.resourceExists(path)).thenReturn(true);
+        objUnderTest.removeServiceProvider(existingIssuer);
```

Comment:
Same method is called again in the assert for the same existingIssuer. Remove one of the method calls. (And in the ideal case the second call should return false ;) )